Added defaults for jdk and server image maps

ChengyuanZhao opened the pull request:
fixes #87

meltsufin commented on java-runtime-builder-app/src/test/java/com/google/cloud/runtimes/builder/injection/RootModuleTest.java:
```diff
@@ -123,4 +129,25 @@ public void testDefaultSettingsPartialCommandLineGiven() throws IOException {
+   * A check that the default settings are not outdated compared to what is in java.yaml.
+   */
+  @Test
+  public void testDefaultSettingsMatchJavaYaml() throws URISyntaxException, IOException {
```
Nice!!

meltsufin commented on java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/Application.java:
```diff
@@ -50,16 +51,37 @@
+      "openjdk9", "gcr.io/google-appengine/openjdk:9"
+
+  );
+  private static final ImmutableMap<String, String> defaultServerMappings;
```
it's final so `DEFAULT_SERVER_MAPPINGS`. Also, you can make it public.

meltsufin commented on java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/Application.java:
```diff
@@ -95,6 +117,14 @@
     addOverrideSettingsToOptions(CLI_OPTIONS);
   }
 
+  public static ImmutableMap<String, String> getDefaultJdkMappings() {
```
Don't need getters for constants.

meltsufin commented on java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/Application.java:
```diff
@@ -50,16 +51,37 @@
+      "openjdk9", "gcr.io/google-appengine/openjdk:9"
+
+  );
+  public static final ImmutableMap<String, String> DEFAULT_SERVER_MAPPINGS;
```
Why don't you use `ImmutableMap.of` here as well as above?

ChengyuanZhao replied:
It's too many arguments :(  `of` is only defined for up to like 8 items.

meltsufin replied:
I see. Have you tried `ImmutableMap.builder().put(...)`?

meltsufin commented on java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/BuildPipelineConfigurator.java:
```diff
@@ -108,6 +117,8 @@ private BuildContext configureBuildContext(Path workspaceDir) throws IOException
         ? parseAppYaml(pathToAppYaml.get())
         : new AppYaml();
 
+    appYaml.applyOverrideSettings(this.overrideSettings);
```
It looks like override settings were never actually applied previously without this line.
Currently, if I remove this line all tests still pass. It doesn't seem right.

ChengyuanZhao replied:
Yes previously this line was done inside the parser, which would work if the app.yaml file contained only part of the required settings, but wouldn't run at all if no app.yaml file was given or found at all.

So now it is moved out here. I'll add the unit tests for this too.

balopat commented:
@meltsufin @ChengyuanZhao  
An idea before we go forward with this approach, please correct me if I'm misunderstanding something: 

The user has the ability to build an application image based on one of our runtimes with just defining an "runtime app config": runtime: java, etc. That maps to one of our digests. This mapping needs to be defined.

1.) In case of `gcloud app deploy` the latest version of the java.yaml, pinned to the latest, approved digests is what's getting pulled in from gs://runtime-builders... 

2.) In case of a local container build - this is what we are trying to achieve here in my understanding - the options are: 
 a.)  (current PR) we  create a set of default mappings from "runtime aliases" to "runtime image digests" in code - but then we have to keep java.yaml in sync with the "default mappings code".
 b.) we let the user pass in a java.yaml (already works) 
 c.) we could use the java.yaml from the runtime-builders bucket by default here as well! 

What's the benefit of a.) vs c.) ?

ChengyuanZhao commented:
@balopat 

Going with c.) would mean the builder application would need to be aware of the structure of java.yaml and would need to parse it. The way it has been (and this pre-dates me, so I'm just going on an assumption) is that the caller of this builder is in charge of parsing java.yaml and needs to be aware of its java.yaml's structure, and provides the settings inside java.yaml to this builder via the command line. To further base the defaults on java.yaml inside this builder would mean we have dependence on java.yaml's parsing and structure at both ends, which seems unclean.

balopat commented:
@ChengyuanZhao Agreed, the dependency on java.yaml is not good - the java.yaml structure is defined by Cloud Container Builder, not the Java Runtime builder. 
The duplication and the test driven sync checking still rubs me the wrong way - this test is just testing for the DEFAULT mappings to be a subset of the (.*)=(.*) regexes, which is fragile, it's a half-way parsing of the yaml: how about spaces and commented lines? 
Another idea: what if we extract the mappings part of the `java.yaml` to a `runtimes-alias-mappings.properties` and use that in the defaults?

ChengyuanZhao commented:
@balopat 
Do you mean we should still get the java.yaml file from the internet and then grab just the section for the mappings? This doesn't feel like it addresses the concerns of java.yaml dependence, though. We would still need to depend on the existence of a java.yaml file and that such a section exists within it. I really think this builder should not be aware of that file at all.

Regarding the existing unit tests just searching for the text within: indeed it is a very simplistic test, but that was the intention. I don't think there should have been a strict contract that the defaults here match those of the Cloud Container Builder, since this runtime builder shouldn't even be aware of the other one's existence. That the defaults in this builder at least appear somewhere in java.yaml (comments or otherwise) is meant to just check that they are not so far out of date that they aren't even mentioned there. 

If we do change our mind and decide that there should be a strict rule that the defaults match, then I would question parsing java.yaml settings and passing them via commandline to this builder in the first place, and would prefer just letting this builder use the java.yaml file directly. However, given the emphasis recently on making this builder easier and easier to use on its own, I think that's a good indication that there shouldn't be a strict contract about these defaults always matching.

balopat commented:
@ChengyuanZhao no, let's forget about java.yaml dependency directly, that is a list of steps for cloud builder and one of the steps happens to have the arguments for runtime mapping. I was suggesting generating the java.yaml from an extracted properties file and using that same properties file to configure the argument defaults as well, that would point to the latest version of the images. 

Question is - is it worth to keep this half-way consistency at all if we still have to remember to keep them in sync anyway? What is the value? This protects against introducing new values in the arg defaults, but if there are new defaults introduced in java.yaml these tests won't fail.

ChengyuanZhao commented:
@balopat @meltsufin 
Please take a look at the most recent changes done from the discussion

meltsufin commented on java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/Application.java:
```diff
@@ -49,56 +50,82 @@
+  public static final String DEFAULT_COMPAT_RUNTIME_IMAGE =
+      "gcr.io/google-appengine/jetty9-compat:latest";
+  public static final String DEFAULT_MAVEN_DOCKER_IMAGE =
+      "gcr.io/cloud-builders/java/mvn:3.5.0-jdk-8";
```
`:latest` maybe?

ChengyuanZhao replied:
Is there a concern with using a fixed version? Now with the unit test if java.yaml gets updated we won't need to worry about this getting out of date.

meltsufin commented on java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/Application.java:
```diff
@@ -49,56 +50,82 @@
+  public static final String DEFAULT_MAVEN_DOCKER_IMAGE =
+      "gcr.io/cloud-builders/java/mvn:3.5.0-jdk-8";
+  public static final String DEFAULT_GRADLE_DOCKER_IMAGE =
+      "gcr.io/cloud-builders/java/gradle:4.0-jdk-8";
```
Alse `:latest`, but we don't have to update `java.yaml` to the same.

ChengyuanZhao replied:
see other reply

meltsufin commented on java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/Application.java:
```diff
@@ -108,7 +135,10 @@ public static void main(String[] args) throws BuildStepException, IOException {
 
     Injector injector = Guice.createInjector(
-        new RootModule(jdkMappings, serverMappings, compatImage, mavenImage, gradleImage,
+        new RootModule(jdkMappings, DEFAULT_JDK_MAPPINGS, serverMappings, DEFAULT_SERVER_MAPPINGS,
```
Why not merge `jdkMappings` with `DEFAULT_JDK_MAPPINGS` here, just like you're doing with `compatImage` and `DEFAULT_COMPAT_RUNTIME_IMAGE`? I think it would be more consistent, and you wouldn't have to make as many changes to the `RootModule`.

ChengyuanZhao replied:
`jdkMappings` is just an array of strings from the command line and there is some more validation that gets done to it in RootModule before it can be merged with the default settings map. We could do it in Application, but that validation function is already in RootModule, so it seems natural to pass both these in.

meltsufin replied:
Would it make sense to just express the defaults also as an array of strings to make them more easily mergable with `jdkMappings`?

ChengyuanZhao replied:
I made the change to merge them in Application. It also solves an issue i had overlooked: previously if a command line setting was formatted different with extra spaces inside we would not match it with the default setting that was actually equivalent and when they get cleaned up (spaces removed) the default setting might actually be kept instead. Now the cleaning up of extra spaces happens before the merging, so that problem is gone.

ChengyuanZhao replied:
PTAL with these changes now up

meltsufin commented:
Thanks @ChengyuanZhao!
There are a lot of changes here. I'll have to come back to review it next week.

meltsufin commented on java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/config/domain/JdkServerLookup.java:
```diff
@@ -71,32 +59,28 @@ private void validate() {
-
-    return jdkKey + KEY_DELIMITER + serverTypeKey;
-  }
+  public abstract Map<String, String> getJdkRuntimeMap();
```
This breaks the encapsulation. You shouldn't need the map if you have the lookup method.
Also, you might want to rethink the the use of an abstract class and instead, if needed, introduce an interface.

ChengyuanZhao replied:
I don't think an interface is necessary, since originally this lookup part was just a single class and it was adequate as that. I did remove these get-map functions in the newest revision.

meltsufin commented on java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/config/domain/JdkServerLookup.java:
```diff
@@ -107,25 +91,36 @@ private String buildServerMapKey(String jdk, String serverType) {
         .collect(Collectors.toList());
   }
 
+  public abstract Map<String, String> getServerRuntimeMap();
```
See comment above on `getJdkRuntimeMap`.

ChengyuanZhao commented:
@meltsufin 
I removed the functions to get the maps of the settings, but I decided to keep them as abstract classes because the validation bit of code can be kept. PTAL

meltsufin commented on java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/Application.java:
```diff
@@ -205,4 +243,68 @@ private static void doForEachOverrideSetting(List<Field> configFields,
+   * @return the merged settings.
+   */
+  @VisibleForTesting
+  public static JdkServerLookup mergeSettingsWithDefaults(String[] rawServerSettings,
```
Is there any reason everything in this method cannot move to the `JdkServerLookup` class?
You could simply merge CLI and default settings before passing them into the `JdkServerLookup` constructor. You just need to make sure to maintain lookup precedence order.
This would eliminate the abstract class and template methods, which should generally be avoided. (see Item 18, in Effective Java 2nd Edition)

ChengyuanZhao replied:
Regarding not having `mergeSettingsWithDefaults` in `JdkServerLookup`:
The reason I didn't put it in there is because this function produces a version of JdkServerLookup that is kinda ad-hoc, and it's actually merging two JdkServerLookup objects to do so. It didn't make sense to me that JdkServerLookup objects would by default have a function to merge two JdkServerLookup objects in that very specific way. 

Regarding merging CLI and Default settings before passing to JdkServerLookup:
The reason I specifically avoided doing that is for the text validation stuff. For example, the user might input the same setting that looks very different because of spaces within the text or just one version of a setting might just be invalid. If i were to do this type of space-removing/trimming/validating then I would be replicating what's inside JdkServerLookup. 

Regarding not using abstract classes: I looked up item 18 but I think this book here is the wrong edition? Item 18 for me is about favoring static inner classes over nonstatic ones. However, I think there is a time and place for most java features, and I think in this situation the template method style is more elegant. I don't see what the difference wrt an interface, other than that I have factored out some common functionality into JdkServerLookup as an abstract class (vs it being an interface and having no common logic). I think the way I have it now very nicely ensures that both CLI and default settings are valid settings objects on their own, and that the ad-hoc merging of them in Application.java is also a valid settings object.

ChengyuanZhao replied:
continuing ... 

Given these requirements: 
1. CLI and default settings both need to be validated
2. CLI and default settings both need to be settings lookup objects on their own
3. A merge of the two settings needs to behave exactly the same way as the original CLI and default settings lookup objects

I think the template method style fits really well.

ChengyuanZhao commented:
One of the checks is stuck? I'll close and reopen to get it to re-evaluate.

meltsufin commented:
I think we just need more test coverage to pass the codecov check?
